File: src/Plotly.NET/Traces/ObjectAbstractions/Contours.fs
namespace Plotly.NET.TraceObjects

open Plotly.NET
open Plotly.NET.LayoutObjects
open Plotly.NET.TraceObjects
open DynamicObj
open System
open System.Runtime.InteropServices

type ContourProject() =
    inherit DynamicObj()

    static member init
        (
            ?X: bool,
            ?Y: bool,
            ?Z: bool
        ) =
        ContourProject() |> ContourProject.style (?X = X, ?Y = Y, ?Z = Z)

    static member style
        (
            ?X: bool,
            ?Y: bool,
            ?Z: bool
        ) =

        fun (contourProject: ContourProject) ->

            contourProject
            |> DynObj.withOptionalProperty "x" X
            |> DynObj.withOptionalProperty "y" Y
            |> DynObj.withOptionalProperty "z" Z

/// Contour object inherits from dynamic object
type Contour() =
    inherit DynamicObj()

    /// Initialized a Contour object
    //[<CompiledName("init")>]
    static member init
        (
            ?Color: Color,
            ?End: float,
            ?Highlight: bool,
            ?HighlightColor: Color,
            ?HighlightWidth: float,
            ?Project: ContourProject,
            ?Show: bool,
            ?Size: float,
            ?Start: float,
            ?UseColorMap: bool,
            ?Width: float
        ) =
        Contour()
        |> Contour.style (
            ?Color = Color,
            ?End = End,
            ?Highlight = Highlight,
            ?HighlightColor = HighlightColor,
            ?HighlightWidth = HighlightWidth,
            ?Project = Project,
            ?Show = Show,
            ?Size = Size,
            ?Start = Start,
            ?UseColorMap = UseColorMap,
            ?Width = Width
        )


    // Applies the styles to Contours()
    //[<CompiledName("style")>]
    static member style
        (
            ?Color: Color,
            ?End: float,
            ?Highlight: bool,
            ?HighlightColor: Color,
            ?HighlightWidth: float,
            ?Project: ContourProject,
            ?Show: bool,
            ?Size: float,
            ?Start: float,
            ?UseColorMap: bool,
            ?Width: float
        ) =

        fun (contour: Contour) ->
            contour
            |> DynObj.withOptionalProperty "color" Color
            |> DynObj.withOptionalProperty "end" End
            |> DynObj.withOptionalProperty "highlight" Highlight
            |> DynObj.withOptionalProperty "highlightcolor" HighlightColor
            |> DynObj.withOptionalProperty "highlightwidth" HighlightWidth
            |> DynObj.withOptionalProperty "project" Project
            |> DynObj.withOptionalProperty "show" Show
            |> DynObj.withOptionalProperty "size" Size
            |> DynObj.withOptionalProperty "start" Start
            |> DynObj.withOptionalProperty "usecolormap" UseColorMap
            |> DynObj.withOptionalProperty "width" Width

/// Contours type inherits from dynamic object
type Contours() =
    inherit DynamicObj()

    static member init
        (
            ?Coloring: StyleParam.ContourColoring,
            ?End: float,
            ?LabelFont: Font,
            ?LabelFormat: string,
            ?Operation: StyleParam.ConstraintOperation,
            ?ShowLabels: bool,
            ?ShowLines: bool,
            ?Size: float,
            ?Start: float,
            ?Type: StyleParam.ContourType,
            ?Value: #IConvertible
        ) =
        Contours()
        |> Contours.style (
            ?Coloring = Coloring,
            ?End = End,
            ?LabelFont = LabelFont,
            ?LabelFormat = LabelFormat,
            ?Operation = Operation,
            ?ShowLabels = ShowLabels,
            ?ShowLines = ShowLines,
            ?Size = Size,
            ?Start = Start,
            ?Type = Type,
            ?Value = Value

        )


    /// Initialized Contours object
    //[<CompiledName("init")>]
    static member initSurface
        (
            ?X: Contour,
            ?Y: Contour,
            ?Z: Contour
        ) =
        Contours() |> Contours.style (?X = X, ?Y = Y, ?Z = Z)

    // Applies the styles to Contours()
    //[<CompiledName("style")>]
    static member style
        (
            ?X: Contour,
            ?Y: Contour,
            ?Z: Contour,
            ?Coloring: StyleParam.ContourColoring,
            ?End: float,
            ?LabelFont: Font,
            ?LabelFormat: string,
            ?Operation: StyleParam.ConstraintOperation,
            ?ShowLabels: bool,
            ?ShowLines: bool,
            ?Size: float,
            ?Start: float,
            ?Type: StyleParam.ContourType,
            ?Value: #IConvertible
        ) =

        fun (contours: Contours) ->

            contours
            |> DynObj.withOptionalProperty "x" X
            |> DynObj.withOptionalProperty "y" Y
            |> DynObj.withOptionalProperty "z" Z
            |> DynObj.withOptionalPropertyBy "coloring" Coloring StyleParam.ContourColoring.convert
            |> DynObj.withOptionalProperty "end" End
            |> DynObj.withOptionalProperty "labelfont" LabelFont
            |> DynObj.withOptionalProperty "labelformat" LabelFormat
            |> DynObj.withOptionalPropertyBy "operation" Operation StyleParam.ConstraintOperation.convert
            |> DynObj.withOptionalProperty "showlabels" ShowLabels
            |> DynObj.withOptionalProperty "showlines" ShowLines
            |> DynObj.withOptionalProperty "size" Size
            |> DynObj.withOptionalProperty "start" Start
            |> DynObj.withOptionalPropertyBy "type" Type StyleParam.ContourType.convert
            |> DynObj.withOptionalProperty "value" Value



    // Initialized x-y-z-Contours with the same properties
    static member initXyz
        (
            ?Color: Color,
            ?End: float,
            ?Highlight: bool,
            ?HighlightColor: Color,
            ?HighlightWidth: float,
            ?Project: ContourProject,
            ?Show: bool,
            ?Size: float,
            ?Start: float,
            ?UseColorMap: bool,
            ?Width: float
        ) =
        Contours()
        |> Contours.styleXyz (
            ?Color = Color,
            ?End = End,
            ?Highlight = Highlight,
            ?HighlightColor = HighlightColor,
            ?HighlightWidth = HighlightWidth,
            ?Project = Project,
            ?Show = Show,
            ?Size = Size,
            ?Start = Start,
            ?UseColorMap = UseColorMap,
            ?Width = Width
        )

    // Applies the styles to Contours()
    //[<CompiledName("styleXyz")>]
    static member styleXyz
        (
            ?Color: Color,
            ?End: float,
            ?Highlight: bool,
            ?HighlightColor: Color,
            ?HighlightWidth: float,
            ?Project: ContourProject,
            ?Show: bool,
            ?Size: float,
            ?Start: float,
            ?UseColorMap: bool,
            ?Width: float
        ) =

        (fun (contours: Contours) ->
            let xyzContour =
                Contour.init (
                    ?Show = Show,
                    ?Color = Color,
                    ?UseColorMap = UseColorMap,
                    ?Width = Width,
                    ?Highlight = Highlight,
                    ?HighlightColor = HighlightColor,
                    ?HighlightWidth = HighlightWidth,
                    ?Project = Project
                )

            contours |> Contours.style (X = xyzContour, Y = xyzContour, Z = xyzContour))

File: src/Plotly.NET.CSharp/ObjectAbstractions/TraceObjects/Contours.cs

using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Plotly.NET.CSharp;

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.ContourProject ContourProject
        (

            Optional<bool> X = default,
            Optional<bool> Y = default,
            Optional<bool> Z = default
        ) => 
            Plotly.NET.TraceObjects.ContourProject.init(

                X: X.ToOption(),
                Y: Y.ToOption(),
                Z: Z.ToOption()
            );
}

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.Contour Contour
        (

            Optional<Color> Color = default,
            Optional<double> End = default,
            Optional<bool> Highlight = default,
            Optional<Color> HighlightColor = default,
            Optional<double> HighlightWidth = default,
            Optional<Plotly.NET.TraceObjects.ContourProject> Project = default,
            Optional<bool> Show = default,
            Optional<double> Size = default,
            Optional<double> Start = default,
            Optional<bool> UseColorMap = default,
            Optional<double> Width = default
        ) => 
            Plotly.NET.TraceObjects.Contour.init(

                Color: Color.ToOption(),
                End: End.ToOption(),
                Highlight: Highlight.ToOption(),
                HighlightColor: HighlightColor.ToOption(),
                HighlightWidth: HighlightWidth.ToOption(),
                Project: Project.ToOption(),
                Show: Show.ToOption(),
                Size: Size.ToOption(),
                Start: Start.ToOption(),
                UseColorMap: UseColorMap.ToOption(),
                Width: Width.ToOption()
            );
}

public static partial class TraceObjects
{
    public static Plotly.NET.TraceObjects.Contours Contours<ValueType>
        (

            Optional<StyleParam.ContourColoring> Coloring = default,
            Optional<double> End = default,
            Optional<Font> LabelFont = default,
            Optional<string> LabelFormat = default,
            Optional<StyleParam.ConstraintOperation> Operation = default,
            Optional<bool> ShowLabels = default,
            Optional<bool> ShowLines = default,
            Optional<double> Size = default,
            Optional<double> Start = default,
            Optional<StyleParam.ContourType> Type = default,
            Optional<ValueType> Value = default
        ) 
            where ValueType : IConvertible
            => 
            Plotly.NET.TraceObjects.Contours.init(

                Coloring: Coloring.ToOption(),
                End: End.ToOption(),
                LabelFont: LabelFont.ToOption(),
                LabelFormat: LabelFormat.ToOption(),
                Operation: Operation.ToOption(),
                ShowLabels: ShowLabels.ToOption(),
                ShowLines: ShowLines.ToOption(),
                Size: Size.ToOption(),
                Start: Start.ToOption(),
                Type: Type.ToOption(),
                Value: Value.ToOption()
            );
}

